app/src/main/java/org/onionshare/android/ShareManager.kt

```diff
@@ -42,132 +40,125 @@ class ShareManager @Inject constructor(
+                        onStarting(zipState, torState)
+                    }.transformWhile { shareUiState ->
+                        emit(shareUiState)
+                        // only continue collecting while we are starting (otherwise would never stop collecting)
+                        shareUiState is ShareUiState.Starting
+                    }.collect { shareUiState ->
+                        LOG.info("New share state: $shareUiState")
+                        _shareState.value = shareUiState
+                        if (shareUiState is ShareUiState.Error) stopOnError(shareUiState)
+                    }
+                    LOG.info("Observer task finished.")
+                }
+                ensureActive()
+                val zipResult = fileTask.await()
+                if (zipResult is ZipResult.Zipped) {
+                    val port = webserverManager.start(zipResult.sendPage)
+                    torManager.publishOnionService(port)
+                    observerTask.await()
+                } else if (zipResult is ZipResult.Error) {
+                    stopOnError(ShareUiState.ErrorAddingFile(zipResult.errorFile))
+                }
             }
         }
     }
 
-    private suspend fun stopSharing() = withContext(Dispatchers.IO) {
-        shouldStop.value = true
+    private fun onStarting(zipState: ZipState?, torState: TorState): ShareUiState {
+        return when (torState) {
+            is TorState.Starting -> {
+                // Tor stays in Starting state as long as the HS descriptor wasn't published.
+                val torPercent = (torState as? TorState.Starting)?.progress ?: 0
+                ShareUiState.Starting(zipState?.progress ?: 0, torPercent)
+            }
+
+            is TorState.Published -> {
+                // We only create the hidden service after files have been zipped and webserver was started,
+                // so we are in sharing state once the first HS descriptor has been published.
+                notificationManager.onSharing()
+                ShareUiState.Sharing(torState.onion)
+            }
+
+            TorState.FailedToConnect -> {
+                ShareUiState.ErrorStarting(true)
+            }
+
+            TorState.Stopped -> {
+                ShareUiState.ErrorStarting(errorMsg = "Tor stopped unexpectedly.")
+            }
+        }
+    }
+
+    private suspend fun stopSharing(errorState: ShareUiState.Error? = null) = withContext(Dispatchers.IO) {
         LOG.info("Stopping sharing...")
+        _shareState.value = ShareUiState.Stopping
         if (startSharingJob?.isActive == true) {
-            // TODO check if this always works as expected
+            LOG.info("Wait for start job to finish...")
             startSharingJob?.cancelAndJoin()
+            LOG.info("Start job to finished.")
         }
         startSharingJob = null
 
         if (torManager.state.value !is TorState.Stopped) torManager.stop()
```

Comment:
Making Onionwrapper more tolerant would be nice. However, it sounds like you want us to bring back the Stopping state again ;)

Reply:
I think the `Stopping` state probably wouldn't solve the problem on its own: we receive the state asynchronously from onionwrapper, so in theory it would be possible to call `stop()` twice before receiving the `Stopping` state from the first call.

So perhaps locally keeping track of the fact that we've called `stop()` might be a better solution? Although I should make onionwrapper more tolerant either way, as this shows that the current behaviour isn't nice for 1/3 apps. :-)